std/zig/parse.zig

```diff
@@ -1,3659 +1,3044 @@
 const std = @import("../std.zig");
 const assert = std.debug.assert;
-const mem = std.mem;
+const Allocator = std.mem.Allocator;
 const ast = std.zig.ast;
-const Tokenizer = std.zig.Tokenizer;
-const Token = std.zig.Token;
+const Node = ast.Node;
+const Tree = ast.Tree;
+const AstError = ast.Error;
 const TokenIndex = ast.TokenIndex;
-const Error = ast.Error;
+const Token = std.zig.Token;
+const TokenIterator = Tree.TokenList.Iterator;
 
-/// Result should be freed with tree.deinit() when there are
-/// no more references to any of the tokens or nodes.
-pub fn parse(allocator: *mem.Allocator, source: []const u8) !*ast.Tree {
+pub const Error = error{UnexpectedToken} || Allocator.Error;
+
+pub fn parse(allocator: *Allocator, source: []const u8) !*Tree {
     var tree_arena = std.heap.ArenaAllocator.init(allocator);
     errdefer tree_arena.deinit();
+    const arena = &tree_arena.allocator;
```

Comment:
This should be refactored similar to https://github.com/ziglang/zig/pull/2424/commits/cb4fb14b6e66bd213575f69eec9598be8394fae6

Reply:
Note to self: don't copy `tree_arena`

Reply:
Yeah there's a bit of a footgun here - sorry about that. The in-progress copy elision changes + "fixed" fields should be able to address this.